working/2509-private-named-parameters/feature-specification.md

```diff
@@ -0,0 +1,382 @@
+    the class is in application code where it's not imported anyway so it
+    doesn't matter much.)
+
+*   They can make the parameter positional instead.
+
+*   They can commit to the API they want by making the field public and the
+    parameter named, but use a public name for the parameter. Then instead of
+    using an initializing formal, they manually initialize the field from the
+    parameter, as in:
+
+    ```dart
+    class House {
+      int? _windows;
+      int? _bedrooms;
+      int? _swimmingPools;
+
+      House({
+        int? windows,
+        int? bedrooms,
+        int? swimmingPools,
+      })  : _windows = windows,
+            _bedrooms = bedrooms,
+            _swimmingPools = swimmingPools;
+    }
+    ```
+
+The first two are hard to measure without interviewing users because we don't
+know which fields are intentionally public and which constructor parameters are
+intentionally positional versus which are simply workarounds to the above
+problem.
+
+We can get data on the third one. I scanned a large corpus of pub packages and
+looked at every constructor initializer:
+
+```
+-- Parameter initializer (42184 total) --
+  32958 ( 78.129%): Other                     ===========================
+   9226 ( 21.871%): Make private: _foo = foo  ========
+```
+
+Over a fifth of all field initializers are simply initializing a private a field
+with a parameter whose name is the same as the field with the underscore
+stripped off. These are places where the user could be using the shorter `this.`
+initializing formal if the underscore wasn't getting in the way.
+
+### Primary constructors
+
+The language team is currently working on a proposal for [primary
+constructors][]. That proposal lets a constructor parameter not just
+*initialize* a field, but *declare* it too.
+
+[primary constructors]: https://github.com/dart-lang/language/blob/main/working/2364%20-%20primary%20constructors/feature-specification.md
+
+This is a highly desired feature. The most-voted open issue in the language repo
+is for [data classes][]. That concept encompasses a few features, but if you
+look at the comments, many users are specifically requesting primary
+constructors. The issue for [primary constructors][primary ctors issue]
+specifically is the #11 upvoted request.
+
+[data classes]: https://github.com/dart-lang/language/issues/314
+[primary ctors issue]: https://github.com/dart-lang/language/issues/2364
+
+With that feature, users will encounter the limitation around private named
+parameters even more often. In-header primary constructors exacerbate this issue
+because an in-header primary constructor *can't* have an initializer list. Users
+will be forced to either make the field public, the parameter positional, or not
+use a primary constructor at all.
+
+We don't want to put users in a position where the code that's most pleasant to
+write requires them to sacrifice semantic properties they want like
+encapsulation.
+
```

Comment:
At this point it would make sense to add a short paragraph to mention that this proposal contains a couple of rules that are concerned with primary constructors (just so the reader doesn't get confused when this proposal talks about a 'field parameter').

Reply:
OK, reading it again I can see that line 267 says '(for a primary constructor)', but I didn't notice that on the first reading.